### PR TITLE
new ray-mesh intersection methods

### PR DESCRIPTION
**Pull request by dassaf4**
Port ray-mesh/polygon/triangle intersection methods and support classes from native:

- `BarycentricTriangle.intersectRay3d` and `BarycentricTriangle.intersectSegment` return a `TriangleLocationDetail`
- `PolygonOps.intersectRay3d` returns a `PolygonLocationDetail`
- `PolyfaceQuery.intersectRay3d` returns a `FacetLocationDetail` implemented by one of `TriangularFacetLocationDetail`, `ConvexFacetLocationDetail`, or `NonConvexFacetLocationDetail`

Also:
- add support for interpolation of auxiliary mesh data at the vertices of a convex facet
- flesh out the functionality of `BarycentricTriangle`
- new _non-abstract_ support methods in abstract classes `IndexedXYCollection` and `IndexedXYZCollection` with default implementations where possible, otherwise overridden in delivered subclasses.
- fix bug in `XYZ.setAt`
- use recommended type "node" instead of "pwa-node" in launch.json for Geometry tests.

Note: civil team will need this backported to 3.6.x, 3.7.x.

### CODE REVIEW

**Comment by pmconne** [comment]
@Mergifyio backport release/3.6.x

**Comment by pmconne** [comment]
@Mergifyio backport release/3.7.x

File: docs/changehistory/NextVersion.md [hunk 1]
```diff
@@ -37,10 +38,16 @@ The new static method [PolyfaceQuery.cloneOffset]($core-geometry) creates a mesh
...
 
 The image below illustrates results with a more complex cross section.
 
 ![Offset Example 2](./assets/cloneOffsetMeshExample2.png "Offset with sharp corners and with chamfers.")
 
 The lower left is the original (smaller, inside) mesh with the (transparent) offset mesh around it with all sharp corners. At upper right the offset has chamfers, again due to setting the `chamferAngleBetweenNormals` to 120 degrees.
+
+## Mesh intersection with ray
+
+New functionality computes the intersection(s) of a `Ray3d` with a `Polyface`. By default, [PolyfaceQuery.intersectRay3d]($core-geometry) returns a [FacetLocationDetail]($core-geometry) for the first found facet that intersects the infinite line parameterized by the ray. A callback can be specified in the optional [FacetIntersectOptions]($core-geometry) parameter to customize intersection processing, e.g., to filter and collect multiple intersections. Other options control whether to populate the returned detail with interpolated auxiliary vertex data: normals, uv parameters, colors, and/or the barycentric scale factors used to interpolate such data.
```
Comment (pmconne) [inline docs/changehistory/NextVersion.md:51]:
```suggestion
New functionality computes the intersection(s) of a [Ray3d]($core-geometry) with a [Polyface]($core-geometry). By default, [PolyfaceQuery.intersectRay3d]($core-geometry) returns a [FacetLocationDetail]($core-geometry) for the first found facet that intersects the infinite line parameterized by the ray. A callback can be specified in the optional [FacetIntersectOptions]($core-geometry) parameter to customize intersection processing, e.g., to filter and collect multiple intersections. Other options control whether to populate the returned detail with interpolated auxiliary vertex data: normals, uv parameters, colors, and/or the barycentric scale factors used to interpolate such data.
```

**Comment by aruniverse** [comment]
> Note: civil team will need this backported to 3.6.x, 

we should *not* be backporting new features as a patch, 3.7 will be out _relatively_ soon, they cant wait?

cc @a-gagnon @bsy-nicholasw @simnorm

**Review by pmconne — APPROVED** [review]

**Comment by dassaf4** [comment]
> we should _not_ be backporting new features as a patch, 3.7 will be out _relatively_ soon, they cant wait?

I think this request happened before a 3.7 release was planned...

**Comment by simnorm** [comment]
thanks for this work, 3.7 should be ok. We just don't want this to be on 4.0.

**Comment by dassaf4** [comment]
Ok, great. Someone please cancel the 3.6.x mergifyio backport.

**Review by saeeedtorabi — APPROVED** [review]

**Pull request merged**